lib/dal/src/job/definition/action.rs

```diff
@@ -121,99 +118,185 @@ impl TryFrom<JobInfo> for ActionJob {
 
 #[instrument(
     name = "action_job.action_task",
-    parent = &parent_span,
     skip_all,
     level = "info",
     fields(
-        ?id,
+        si.action.id = ?action_id,
         si.action.kind = Empty,
         si.component.id = Empty,
+        // TODO: determine what this field is called for retries
+        si.action_job.process.retries = Empty,
     )
 )]
-async fn action_task(
+async fn inner_run(
     ctx: &mut DalContext,
-    id: ActionId,
-    parent_span: Span,
+    action_id: ActionId,
 ) -> JobConsumerResult<Option<ActionRunResultSuccess>> {
+    let (prototype_id, component_id) = prepare_for_execution(ctx, action_id).await?;
+
+    // Execute the action function
+    let maybe_resource = ActionPrototype::run(ctx, prototype_id, component_id).await?;
+
+    // Retry process_and_record_execution on a conflict error up to a max
+    retry_on_conflicts(
+        || process_and_record_execution(ctx.clone(), maybe_resource.as_ref(), action_id),
+        10,
+        Duration::from_millis(10),
+        "si.action_job.process.retries",
+    )
+    .await?;
+
+    Ok(maybe_resource)
+}
+
+async fn prepare_for_execution(
+    ctx: &mut DalContext,
+    action_id: ActionId,
+) -> JobConsumerResult<(ActionPrototypeId, ComponentId)> {
     let span = Span::current();
-    let component_id = Action::component_id(ctx, id)
+
+    let component_id = Action::component_id(ctx, action_id)
         .await?
-        .ok_or(ActionError::ComponentNotFoundForAction(id))?;
+        .ok_or(ActionError::ComponentNotFoundForAction(action_id))?;
 
-    let prototype_id = Action::prototype_id(ctx, id).await?;
+    let prototype_id = Action::prototype_id(ctx, action_id).await?;
     let prototype = ActionPrototype::get_by_id(ctx, prototype_id).await?;
     span.record("si.action.kind", &tracing::field::debug(&prototype.kind));
     span.record("si.component.id", &tracing::field::debug(&component_id));
-    Action::set_state(ctx, id, ActionState::Running).await?;
+    Action::set_state(ctx, action_id, ActionState::Running).await?;
 
     // Updates the action's state
     ctx.commit().await?;
     ctx.update_snapshot_to_visibility().await?;
 
-    let status = Action::run(ctx, id).await?;
+    let component_id = Action::component_id(ctx, action_id)
+        .await?
+        .ok_or(ActionError::ComponentNotFoundForAction(action_id))?;
+
+    Ok((prototype_id, component_id))
+}
+
+async fn process_and_record_execution(
```

Comment:
this encapsulates all of the mutations that need to be made after a Func successfully returns and is therefore what we retry if there are conflicts. Given the scope of what we do after a func returns here (removing action node, updating resource, potentially removing the component from the graph) - we should monitor how often we end up retrying so we aren't hiding a false positive conflict case